### PR TITLE
 Add priority class syncer 

### PR DESCRIPTION
**Pull request by weiling61**
Add priority class syncer (priority class is an extra resource, needs to be activated using --extra-syncing-resources flag)

https://github.com/kubernetes-sigs/multi-tenancy/issues/1007

### CODE REVIEW

File: incubator/virtualcluster/pkg/syncer/resources/register.go [hunk 1]
```diff
@@ -58,6 +60,16 @@ func init() {
 	}
 }
 
+// extraResourceToController maps extra resource name with extra resource controller
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/resources/register.go:63]: how about creating a map for these controller?  controller-name -> func
check if enable this controller by `ExtraSyncingResources`
just like kcm https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L351

File: incubator/virtualcluster/pkg/syncer/conversion/equality.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 	v1storage "k8s.io/api/storage/v1"
+	v1priorityclass "k8s.io/api/scheduling/v1"
```
[the remark this answers is not in the document]
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/conversion/equality.go:24]: ok

File: incubator/virtualcluster/pkg/syncer/conversion/helper.go [hunk 3]
```diff
@@ -28,6 +28,7 @@ import (
 	"github.com/pkg/errors"
 	v1 "k8s.io/api/core/v1"
 	storagev1 "k8s.io/api/storage/v1"
+	priorityclassv1 "k8s.io/api/scheduling/v1"
```
[the remark this answers is not in the document]
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/conversion/helper.go:31]: sure

File: incubator/virtualcluster/pkg/syncer/resources/priorityclass/checker.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/resources/priorityclass/checker.go:2]: ok

File: incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+		Provisioner: "p1",
+	}
+
+	for _, f := range mFuncs {
+		f(sc)
+	}
+	return sc
+}
+
+func TestUWPVCreation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go:58]: ok

File: incubator/virtualcluster/pkg/syncer/resources/register.go [hunk 6]
```diff
@@ -72,5 +84,17 @@ func Register(config *config.SyncerConfiguration,
...
+		klog.V(4).Infof("extra resource controllers that will be synced to virtual cluster are %v", r )
+		extraf := extraResourceToController(r)
+		if extraf != nil {
+			if c, _, _, err := extraf(config, client, informerFactory, vcClient, vcInformer, nil); err != nil {
+				klog.Errorf("cannot add extra resource %v for syncer", r)
+				return err
+			} else {
+				controllerManager.AddResourceSyncer(c)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/resources/register.go:97]: done

File: incubator/virtualcluster/pkg/syncer/resources/register.go (hunk 1 above)
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/resources/register.go:63]: sure

File: incubator/virtualcluster/pkg/syncer/mccontroller/mccontroller.go [hunk 7]
```diff
@@ -24,6 +24,7 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 	storagev1 "k8s.io/api/storage/v1"
+	priorityclassv1 "k8s.io/api/scheduling/v1"
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/mccontroller/mccontroller.go:27]: Do you mind change this one to schedulingv1 as well?

File: incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+	testcases := map[string]struct {
+		ExistingObjectInSuper  []runtime.Object
+		ExistingObjectInTenant []runtime.Object
+		EnqueuedKey            string
+		ExpectedCreatedObject  []string
+		ExpectedError          string
+		ExpectedNoOperation    bool
+	}{
+		"pSC exists but vSC not found": {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go:80]: pSC-> pPC, vSC->vPC.  Same for the other tests

File: incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+	testcases := map[string]struct {
+		ExistingObjectInSuper  []runtime.Object
+		ExistingObjectInTenant []runtime.Object
+		EnqueuedKey            string
+		ExpectedDeletedObject  []string
+		ExpectedError          string
+		ExpectedNoOperation    bool
+	}{
+		"pSC not found, vSC exists": {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go:265]: pSC->pPC, vSC->vPC

File: incubator/virtualcluster/pkg/syncer/util/test/runDWS.go [hunk 10]
```diff
@@ -22,6 +22,7 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 	storageV1 "k8s.io/api/storage/v1"
+	priorityclassV1 "k8s.io/api/scheduling/v1"
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/util/test/runDWS.go:25]: priorityclassV1 -> schedulingV1

Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/util/test/runDWS.go:25]: done

File: incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go (hunk 9 above)
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/resources/priorityclass/uws_test.go:265]: sure

File: incubator/virtualcluster/pkg/syncer/mccontroller/mccontroller.go (hunk 7 above)
Comment (weiling61) [reply incubator/virtualcluster/pkg/syncer/mccontroller/mccontroller.go:27]: yes

**Pull request merged**